### PR TITLE
Immutable peer selection refactoring and enhancements

### PR DESCRIPTION
**Pull request by markberger**
This pull request is associated with [ticket #1382](https://tahoe-lafs.org/trac/tahoe-lafs/ticket/1382).

### CODE REVIEW

File: src/allmydata/immutable/upload.py [hunk 1]
```diff
@@ -361,10 +447,27 @@ def _handle_existing_response(self, res, tracker):
                     % (tracker.get_name(), tuple(sorted(buckets))),
                     level=log.NOISY)
             for bucket in buckets:
+                self.peer_selector.add_peer_with_share(serverid, bucket)
```
Comment (str4d) [inline src/allmydata/immutable/upload.py]: I think this is Step 1.

File: src/allmydata/immutable/happiness_upload.py [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+                    peers.remove(peer)
+
+        existing_mappings = self._calculate_mappings(peers, shares, servermap)
+        existing_peers, existing_shares = self._extract_ids(existing_mappings)
+
+        # Calculate share placement for the remaining peers and shares which
+        # won't be preserved by existing allocations.
+        peers = self.peers - existing_peers - used_peers
+        shares = self.shares - existing_shares - used_shares
+        new_mappings = self._calculate_mappings(peers, shares)
```
Comment (str4d) [inline src/allmydata/immutable/happiness_upload.py]: Step 7 (and possibly 6?)

File: src/allmydata/immutable/happiness_upload.py [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+    def generate_mappings(self):
+        """
+        Generates the allocations the upload should based on the given
+        information. We construct a dictionary of 'share_num' -> set(server_ids)
+        and return it to the caller. Each share should be placed on each server
+        in the corresponding set. Existing allocations appear as placements
+        because attempting to place an existing allocation will renew the share.
+        """
+
+        # First calculate share placement for the readonly servers.
```
Comment (str4d) [inline src/allmydata/immutable/happiness_upload.py]: Step 2.

File: src/allmydata/immutable/happiness_upload.py [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+        for peer in self.servermap:
+            if peer in self.readonly_peers:
+                readonly_map.setdefault(peer, self.servermap[peer])
+                for share in self.servermap[peer]:
+                    readonly_shares.add(share)
+
+        readonly_mappings = self._calculate_mappings(readonly_peers, readonly_shares, readonly_map)
+        used_peers, used_shares = self._extract_ids(readonly_mappings)
+
+        # Calculate share placement for the remaining existing allocations
```
Comment (str4d) [inline src/allmydata/immutable/happiness_upload.py]: Step 4.

File: src/allmydata/immutable/happiness_upload.py [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+        readonly_peers = self.readonly_peers
+        readonly_shares = set()
+        readonly_map = {}
+        for peer in self.servermap:
+            if peer in self.readonly_peers:
+                readonly_map.setdefault(peer, self.servermap[peer])
+                for share in self.servermap[peer]:
+                    readonly_shares.add(share)
+
+        readonly_mappings = self._calculate_mappings(readonly_peers, readonly_shares, readonly_map)
```
Comment (str4d) [inline src/allmydata/immutable/happiness_upload.py]: Step 3.

File: src/allmydata/immutable/happiness_upload.py (hunk 4 above)
Comment (str4d) [inline src/allmydata/immutable/happiness_upload.py]: Step 4.

File: src/allmydata/immutable/happiness_upload.py [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+        for peer in self.servermap:
+            if peer in used_peers:
+                servermap.pop(peer, None)
+            else:
+                servermap[peer] = servermap[peer] - used_shares
+                if servermap[peer] == set():
+                    servermap.pop(peer, None)
+                    peers.remove(peer)
+
+        existing_mappings = self._calculate_mappings(peers, shares, servermap)
```
Comment (str4d) [inline src/allmydata/immutable/happiness_upload.py]: Step 5.

File: src/allmydata/immutable/happiness_upload.py [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+        existing_peers, existing_shares = self._extract_ids(existing_mappings)
+
+        # Calculate share placement for the remaining peers and shares which
+        # won't be preserved by existing allocations.
```
Comment (str4d) [inline src/allmydata/immutable/happiness_upload.py]: Steps 6 and 7 (I think in aggregate up to the end of the method).

File: src/allmydata/immutable/upload.py [hunk 8]
```diff
@@ -386,130 +489,61 @@ def _get_progress_message(self):
...
+        if self._status:
+            self._status.set_status("Contacting Servers [%s] (first query),"
+                                    " %d shares left.."
+                                    % (tracker.get_name(),
+                                       len(self.homeless_shares)))
+        return (tracker, shares_to_ask)
+
+
+    def _request_another_allocation(self):
+        allocation = self._get_next_allocation()
```
Comment (str4d) [inline src/allmydata/immutable/upload.py]: Steps 8 and 9.

File: src/allmydata/immutable/upload.py [hunk 9]
```diff
@@ -553,6 +592,7 @@ def _got_response(self, res, tracker, shares_to_ask, put_tracker_here):
                     level=log.NOISY)
             progress = False
             for s in alreadygot:
+                self.peer_selector.confirm_share_allocation(s, tracker.get_serverid())
```
Comment (str4d) [inline src/allmydata/immutable/upload.py]: This appears to do the inverse of Step 10, but I haven't located Step 10 itself yet.

File: src/allmydata/immutable/upload.py [hunk 10]
```diff
@@ -565,6 +605,8 @@ def _got_response(self, res, tracker, shares_to_ask, put_tracker_here):
             if allocated:
                 self.use_trackers.add(tracker)
                 progress = True
+                for s in allocated:
+                    self.peer_selector.confirm_share_allocation(s, tracker.get_serverid())
```
Comment (str4d) [inline src/allmydata/immutable/upload.py]: Also inverse of Step 10.

File: src/allmydata/immutable/upload.py [hunk 11]
```diff
@@ -351,6 +436,7 @@ def _handle_existing_response(self, res, tracker):
         if isinstance(res, failure.Failure):
             self.log("%s got error during existing shares check: %s"
                     % (tracker.get_name(), res), level=log.UNUSUAL)
+            self.peer_selector.mark_bad_peer(serverid)
```
Comment (str4d) [inline src/allmydata/immutable/upload.py]: Maybe part of Step 10, but for existing shares.

File: src/allmydata/immutable/upload.py [hunk 12]
```diff
@@ -361,10 +447,27 @@ def _handle_existing_response(self, res, tracker):
...
-            self.full_count += 1
-            self.bad_query_count += 1
+
+    def _handle_existing_write_response(self, res, tracker, shares_to_ask):
+        """
+        Function handles the response from the write servers
+        when inquiring about what shares each server already has.
+        """
+        if isinstance(res, failure.Failure):
+            self.peer_selector.mark_bad_peer(tracker.get_serverid())
```
Comment (str4d) [inline src/allmydata/immutable/upload.py]: Also maybe part of Step 10.

File: src/allmydata/test/test_upload.py [hunk 13]
```diff
@@ -632,6 +612,21 @@ def _check(res):
...
+        # This tests ensures that Tahoe only contacts 2n servers
+        # during peer selection
+        self.make_client(40)
+        self.set_encoding_parameters(3, 7, 10)
+        data = self.get_data(SIZE_LARGE)
+        d = upload_data(self.u, data)
+        def _check(res):
+            servers_contacted = []
+            for s in self.node.last_servers:
+                if(s.queries != 0):
```
Comment (daira) [inline src/allmydata/test/test_upload.py]: Nit: ``if s.queries != 0:``

**Comment by meejah** [comment]
This actually landed via https://github.com/tahoe-lafs/tahoe-lafs/pull/416

**Pull request closed without merging**